### PR TITLE
Add recycle bin operations #33

### PR DESCRIPTION
**Pull request by ChiuchiuSorin**
Create a hyperlink inside the Health Check section that allows the user to view all wikis recycle bins and perform actions over them.

### CODE REVIEW

**Comment by ChiuchiuSorin** [comment]
* Health check shrink recycle bin modal: 
image

* Wiki recycle bin page: 
image

There is the same view for deleted attachments, but it's missing the `Delete all` button.

**Comment by ChiuchiuSorin** [comment]
There is no longer a dedicated page inside Admin tools to view the recycle bin. The modal now has hyperlinks for each wiki to Page Index document, with a specific tab being preselected depending on the category (pages or attachments).
image

File: application-admintools-api/src/main/java/com/xwiki/admintools/health/WikiRecycleBinResult.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    public void setAttachmentSize(long attachmentSize)
+    {
+        this.attachmentSize = attachmentSize;
+    }
+
+    /**
+     * Get the number of deleted documents in this wiki.
+     * @return number of deleted documents in wiki.
+     */
+    public long getPageSize()
```
Comment (oanalavinia) [inline application-admintools-api/src/main/java/com/xwiki/admintools/health/WikiRecycleBinResult.java:70]: WDYT about `getDocumentsCount`?

File: application-admintools-api/src/main/java/com/xwiki/admintools/health/WikiRecycleBinResult.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public WikiRecycleBinResult()
+    {
+    }
+
+    /**
+     * Get the number of deleted attachments in this wiki.
+     * @return number of deleted attachments in wiki.
+     */
+    public long getAttachmentSize()
```
Comment (oanalavinia) [inline application-admintools-api/src/main/java/com/xwiki/admintools/health/WikiRecycleBinResult.java:52]: WDYT about `getAttachmentsCount`?

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/health/operations/RecycleBinOperations.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import com.xwiki.admintools.health.WikiRecycleBinResult;
+
+/**
+ * Retrieve data about wikis recycle bins.
+ *
+ * @version $Id$
+ */
+@Component(roles = RecycleBinOperations.class)
+@Singleton
+public class RecycleBinOperations
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/health/operations/RecycleBinOperations.java:46]: `RecycleBinsManager` WDYT?

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/AdminToolsManager.java [hunk 4]
```diff
@@ -148,4 +156,19 @@ public List<String> getPagesOverGivenNumberOfComments(long maxComments) throws Q
...
+     * Return a {@link List} of {@link WikiRecycleBinResult} that is populated with results for all existing wikis in
+     * instance.
+     *
+     * @return info about all existing wikis in instance.
+     * @throws QueryException when there is an issue regarding the queries that retrieve the number of deleted
+     *     documents and attachments.
+     * @throws WikiManagerException for any exception while retrieving the {@link Collection} of
+     *     {@link WikiDescriptor}.
+     */
+    public List<WikiRecycleBinResult> getWikisRecycleBinSize() throws QueryException, WikiManagerException
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/AdminToolsManager.java:170]: `Bins` should be used

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/health/operations/RecycleBinOperations.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     * Generate a {@link List} of {@link WikiRecycleBinResult} that is populated with results for all existing wikis in
+     * instance.
+     *
+     * @return info about all existing wikis in instance.
+     * @throws QueryException when there is an issue regarding the queries that retrieve the number of deleted
+     *     documents and attachments.
+     * @throws WikiManagerException for any exception while retrieving the {@link Collection} of
+     *     {@link WikiDescriptor}.
+     */
+    public List<WikiRecycleBinResult> getAllWikisRecycleBinInfo() throws QueryException, WikiManagerException
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/health/operations/RecycleBinOperations.java:64]: Maybe `getWikisRecycleBins`? Or `getWikisRecycleBinsSize`, to have it similar to the method where is used

File: application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml [hunk 6]
```diff
@@ -228,5 +229,64 @@
...
+                    'adminTools.dashboard.healthcheck.modal.wikiBins.header.wiki'))&lt;/th&gt;
+                  &lt;th&gt;$escapetool.xml($services.localization.render(
+                    'adminTools.dashboard.healthcheck.modal.wikiBins.header.page'))&lt;/th&gt;
+                  &lt;th&gt;$escapetool.xml($services.localization.render(
+                    'adminTools.dashboard.healthcheck.modal.wikiBins.header.attach'))&lt;/th&gt;
+                  &lt;th&gt;$escapetool.xml($services.localization.render(
+                    'adminTools.dashboard.healthcheck.modal.wikiBins.header.total'))&lt;/th&gt;
+                &lt;/tr&gt;
+                #foreach($wikiBin in $wikisRecycleBin)
+                  #set ($wikiRecycleReference = "${wikiBin.getWikiId()}:Main.AllDocs")
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml:261]: `#set ($allDocsRef = $services.model.createDocumentReference('wikiId', 'Main', 'AllDocs'))`

File: application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml [hunk 7]
```diff
@@ -228,5 +229,64 @@
...
+                  &lt;tr&gt;
+                    &lt;td&gt;$wikiBin.getWikiName()&lt;/td&gt;
+                    &lt;td&gt;
+                      &lt;a href="${xwiki.getURL($wikiRecycleReference)}?view=deletedDocs" target="_blank"&gt;
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml:265]:
```
${xwiki.getURL($allDocsRef, 'view', $escapetool.url({'view': 'deletedDocs'}))}
```

File: application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml [hunk 8]
```diff
@@ -228,5 +229,64 @@
...
+                        $wikiBin.getPageSize()&lt;/a&gt;
+                    &lt;/td&gt;
+                    &lt;td&gt;
+                      &lt;a href="${xwiki.getURL($wikiRecycleReference)}?view=deletedAttachments" target="_blank"&gt;
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml:269]: Same

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 9]
```diff
@@ -191,8 +192,15 @@ adminTools.dashboard.spamPage.modal.title=Pages with more than {0} comments
...
 adminTools.dashboard.spamPage.modal.header.noComments=Number of comments
-adminTools.dashboard.healthcheck.modal.content=Are you sure you want to flush the instance cache?
-adminTools.dashboard.healthcheck.modal.button=Confirm</content>
+adminTools.dashboard.healthcheck.modal.content=Are you sure you want to flush XWiki cache?
+adminTools.dashboard.healthcheck.modal.button=Confirm
+adminTools.dashboard.showLogs=View last n lines of log...
+adminTools.dashboard.healthcheck.modal.wikiBins.title=Wikis recycle bin (number of documents)
+adminTools.dashboard.healthcheck.modal.wikiBins.error=There has been an error while gathering info about your instance wikis recycle bin size.
+adminTools.dashboard.healthcheck.modal.wikiBins.header.wiki=Wiki
+adminTools.dashboard.healthcheck.modal.wikiBins.header.page=Page recycle bin
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:201]: `Deleted pages`

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 10]
```diff
@@ -191,8 +192,15 @@ adminTools.dashboard.spamPage.modal.title=Pages with more than {0} comments
...
+adminTools.dashboard.healthcheck.modal.wikiBins.header.attach=Attachment recycle bin
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:202]: `Deleted attachments`

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 11]
```diff
@@ -191,8 +192,15 @@ adminTools.dashboard.spamPage.modal.title=Pages with more than {0} comments
 adminTools.dashboard.spamPage.modal.notFound=There are no pages with more than {0} comments.
 adminTools.dashboard.spamPage.modal.header.page=Page
 adminTools.dashboard.spamPage.modal.header.noComments=Number of comments
-adminTools.dashboard.healthcheck.modal.content=Are you sure you want to flush the instance cache?
-adminTools.dashboard.healthcheck.modal.button=Confirm</content>
+adminTools.dashboard.healthcheck.modal.content=Are you sure you want to flush XWiki cache?
+adminTools.dashboard.healthcheck.modal.button=Confirm
+adminTools.dashboard.showLogs=View last n lines of log...
+adminTools.dashboard.healthcheck.modal.wikiBins.title=Wikis recycle bin (number of documents)
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:198]: `Recycle bins size for all wikis`

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 12]
```diff
@@ -191,8 +192,15 @@ adminTools.dashboard.spamPage.modal.title=Pages with more than {0} comments
...
+adminTools.dashboard.healthcheck.modal.wikiBins.error=There has been an error while gathering info about your instance wikis recycle bin size.
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:199]: `There has been an error while gathering recycle bins size info for all wikis.`

File: application-admintools-default/src/main/resources/templates/wikiSizeTemplate.vm [hunk 13]
```diff
@@ -57,9 +57,12 @@
           <li><a href="#pagesOverNumberOfComments" data-toggle="modal" data-target="#pagesOverNumberOfComments">
             $escapetool.xml($services.localization.render('adminTools.dashboard.instanceUsage.specific.comments',
             [$spamLimit]))</a></li>
+          <li><a href="#viewWikisRecycleBinModal" data-toggle="modal" data-target="#viewWikisRecycleBinModal">
```
Comment (oanalavinia) [inline application-admintools-default/src/main/resources/templates/wikiSizeTemplate.vm:60]: You are also missing a title for this link

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-admintools-default/src/main/resources/templates/wikiSizeTemplate.vm (hunk 13 above)
Comment (oanalavinia) [reply application-admintools-default/src/main/resources/templates/wikiSizeTemplate.vm]: You should add a title to this element, so when the user hovers over it he will get more info about what it does

**Comment by ChiuchiuSorin** [comment]
The look of the modal with the explanatory text. 
image

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/wikiUsage/RecycleBinsManager.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        Collection<WikiDescriptor> wikiDescriptors = wikiDescriptorManager.getAll();
+        List<WikiRecycleBins> results = new ArrayList<>();
+        for (WikiDescriptor wikiDescriptor : wikiDescriptors) {
+            String wikiId = wikiDescriptor.getId();
+            WikiRecycleBins result = new WikiRecycleBins();
+            result.setWikiName(wikiDescriptor.getPrettyName());
+            result.setWikiId(wikiId);
+            result.setAttachmentsCount(getNumberOfDeletedDocuments(wikiId, "DeletedAttachment"));
+            result.setDocumentsCount(getNumberOfDeletedDocuments(wikiId, "XWikiDeletedDocument"));
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/wikiUsage/RecycleBinsManager.java:74]: This is just more like a suggestion
You could add a `getWikiRecycleBinsSize` method to create the `WikiRecycleBins` object for this wiki, to separate things. Then you can have here only `results.add(getWikiRecycleBinsSize(wikiDescriptor.getId))`
To go even further, you can write this `for` as:
```
wikiDescriptorManager.getAll().stream()
  .forEach(wiki -> results.add(getWikiRecycleBinsSize(wiki.getId)));
```
Might need some adjustments

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/wikiUsage/RecycleBinsManager.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    @Inject
+    private QueryManager queryManager;
+
+    /**
+     * Generate a {@link List} of {@link WikiRecycleBins} that is populated with results for all existing wikis in
+     * instance.
+     *
+     * @return info about all existing wikis in instance.
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/wikiUsage/RecycleBinsManager.java:58]: `Get instance recycle bins info, like deleted documents and attachment.`

`@return a {@link List} of {@link WikiRecycleBins} objects containing recycle bins info for each wiki of the instance`

File: application-admintools-default/src/main/java/com/xwiki/admintools/internal/AdminToolsManager.java [hunk 16]
```diff
@@ -148,4 +156,19 @@ public List<String> getPagesOverGivenNumberOfComments(long maxComments) throws Q
     {
         return instanceUsage.getDocumentsOverGivenNumberOfComments(maxComments);
     }
+
+    /**
+     * Return a {@link List} of {@link WikiRecycleBins} that is populated with results for all existing wikis in
+     * instance.
+     *
+     * @return info about all existing wikis in instance.
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/internal/AdminToolsManager.java:164]: This needs an update as well, check comment on RecycleBinsManager

File: application-admintools-default/src/main/java/com/xwiki/admintools/script/AdminToolsScriptService.java [hunk 17]
```diff
@@ -179,4 +181,16 @@ public List<String> getHealthCheckJobId() throws AccessDeniedException
         String currentWiki = modelContext.getCurrentEntityReference().extractReference(EntityType.WIKI).getName();
         return List.of("adminTools", "healthCheck", currentWiki);
     }
+
+    /**
+     * Get recycle bin info for all wikis in your instance.
+     *
+     * @return a {@link List} with wikis recycle bin info or null in case of an error.
```
Comment (oanalavinia) [inline application-admintools-default/src/main/java/com/xwiki/admintools/script/AdminToolsScriptService.java:188]: This needs an update as well, check comment on RecycleBinsManager

File: application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml [hunk 18]
```diff
@@ -192,7 +193,16 @@ adminTools.dashboard.spamPage.modal.notFound=There are no pages with more than {
 adminTools.dashboard.spamPage.modal.header.page=Page
 adminTools.dashboard.spamPage.modal.header.noComments=Number of comments
 adminTools.dashboard.healthcheck.modal.content=Are you sure you want to flush the instance cache?
-adminTools.dashboard.healthcheck.modal.button=Confirm</content>
+adminTools.dashboard.healthcheck.modal.button=Confirm
+adminTools.dashboard.showLogs=View last n lines of log...
```
Comment (oanalavinia) [inline application-admintools-ui/src/main/resources/AdminTools/Code/Translations.xml:197]: Duplicated key

**Review by oanalavinia — COMMENTED** [review]
Almost done:)

**Review by oanalavinia — APPROVED** [review]

**Pull request merged**